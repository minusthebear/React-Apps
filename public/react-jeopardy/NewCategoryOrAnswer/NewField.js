import Formsy from "formsy-react";
import BasicTextField from "../common/BasicTextField";
import React, {useState, useEffect} from "react";
import Select from "react-select";
import {getEntriesNames, setValueAndLabel} from './NewFieldService'


const NewField = ({allQuestionData}) => {

    // TODO: change later
    let [keys, setKeys] = useState([]);
    let [cats, setCats] = useState([]);
    let [catFlag, setCatFlag] = useState(false);
    let [values, setValues] = useState(null);
    let [valueFlag, setValueFlag] = useState(false);

    useEffect(() => {
        let arr = Object.keys(allQuestionData);

        setKeys(arr.map((k) => {
            return setValueAndLabel(k);
        }));

    }, [allQuestionData]);

    const selectValues = (val) => {
        const key = val.value;
        if (!allQuestionData[key]) {
            return;
        }

        const arrEntries = getEntriesNames(allQuestionData, key);

        if (!arrEntries.length) {
            return;
        } else if (arrEntries.length === 1) {
            console.log(setValueAndLabel(arrEntries[0]));
            setValues(setValueAndLabel(arrEntries[0]));
            setValueFlag(true);
        } else {
            console.log(arrEntries.map((k) => setValueAndLabel(k)));
            setCats(arrEntries.map((k) => setValueAndLabel(k)))
            setCatFlag(true);
        }


    };

    return (
        <Formsy data-testid="formsy" className="category-number-form" onSubmit={() => {
        }}> <div data-testid="keys" style={{width: '200px', textAlign:'center', margin: '0 auto'}}>
                    <Select options={keys} onChange={selectValues}/>
            </div>
            { catFlag ?
                <div data-testid="categories" style={{width: '200px', textAlign:'center', margin: '0 auto'}}>
                    <Select options={cats} onChange={() => {}}/>
                </div>
                : null
            }
            { valueFlag ?
                <div data-testid="values" style={{width: '200px', textAlign:'center', margin: '0 auto'}}>
                    <BasicTextField
                        name="valueName"
                        value={''}
                        field={values.label}
                        onKeyDown={() => {}}
                        onChange={() => {}}
                        disabled={false}
                    />
                </div>
                : null
            }
        </Formsy>
    );
};

export default NewField;