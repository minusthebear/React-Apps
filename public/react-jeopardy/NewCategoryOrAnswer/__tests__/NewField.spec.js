import React from 'react';
import {shallow, mount} from 'enzyme';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { renderHook } from '@testing-library/react-hooks';
import NewField from '../NewField';
import questions from '../../../__mocks__/dummyQuestions';

// NOTE: if jest.mock('react-select') is not at top of page, many unit tests will fail
jest.mock('react-select', () => (props) => <select>Select</select>);

describe("NewField",function() {

    let newField;

    beforeEach(() => {
        newField = mount(<NewField allQuestionData={questions} />)
    });

    it('renders a Formsy element', () => {
        const formsy = newField.find('Formsy');
        expect(formsy.length).toBe(1);
    });

    it('renders a single div', () => {
        const div = newField.find('div');
        expect(div.length).toBe(1);
    });

    it('renders a single select element', () => {
        const select = newField.find('select');
        expect(select.length).toBe(1);
    });
});

describe('rendering the component with the correct props', () => {
    let wrapper;

    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, setState]);

    beforeEach(() => {
        const {result} = renderHook(() => <NewField allQuestionData={questions} />);
        wrapper = result;
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('allQuestionData', () => {
        expect(wrapper.current.props.allQuestionData).toBe(questions);
    })
});

describe("make snapshot", () => {

    it('renders as expected', () => {
        const tree = renderer.create(<NewField allQuestionData={{}} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});


//
// it('smoke test', () => {
//     let mountedNewField = shallow(<NewField />);
// });
