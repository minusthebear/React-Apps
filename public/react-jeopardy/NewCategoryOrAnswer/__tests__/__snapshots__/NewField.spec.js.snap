// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make snapshot renders as expected 1`] = `
<form
  className="category-number-form"
  data-testid="formsy"
  disabled={false}
  onError={[Function]}
  onReset={[Function]}
  onSubmit={[Function]}
>
   
  <div
    data-testid="keys"
    style={
      Object {
        "margin": "0 auto",
        "textAlign": "center",
        "width": "200px",
      }
    }
  >
    <select>
      Select
    </select>
  </div>
</form>
`;
